File: 2024/d05/src/main.py
from rich import print
from time import time
from functools import cmp_to_key


def timeit(fn):
    def wrapper(*args, **kwargs):
        start = time()
        result = fn(*args, **kwargs)
        elapsed = (time() - start) / 1000
        print(f"{fn.__name__} {elapsed=:.10f}ms {result=}")
        return result

    return wrapper


input = open("src/input").read().strip()

t1, t2 = [text.splitlines() for text in input.split("\n\n")]
order = [tuple(map(int, line.split("|"))) for line in t1]
updates = [tuple(map(int, line.split(","))) for line in t2]


@timeit
def part1():
    out = 0
    for u in updates:
        if not any(a in u and b in u and u.index(a) > u.index(b) for a, b in order):
            out += u[len(u) // 2]
    return out


@timeit
def part1_with_cache():
    cache = {}
    for x, y in order:
        cache[(x, y)] = True
        cache[(y, x)] = False

    out = 0
    for u in updates:
        if not any(k in cache and not cache[k] for k in zip(u, u[1:])):
            out += u[len(u) // 2]
    return out


@timeit
def part2():
    cache = {}
    for x, y in order:
        cache[(x, y)] = -1
        cache[(y, x)] = 1

    out = 0
    for u in updates:
        if any(cache.get(k, 0) == 1 for k in zip(u, u[1:])):
            cmp = cmp_to_key(lambda x, y: cache.get((x, y), 0))
            out += sorted(u, key=cmp)[len(u) // 2]
    return out


part1()
part1_with_cache()
part2()

File: 2024/d05/src/main.rs
use std::collections::HashMap;

const INPUT: &str = include_str!("input");

fn part1() -> i32 {
    let (order, updates) = INPUT.split_once("\n\n").unwrap();

    let mut cache = HashMap::new();
    for line in order.lines() {
        let (x, y) = line
            .split_once("|")
            .map(|(x, y)| (x.parse::<i32>().unwrap(), y.parse::<i32>().unwrap()))
            .unwrap();
        cache.insert((x, y), true);
        cache.insert((y, x), false);
    }

    let mut out = 0;
    for line in updates.lines() {
        let u: Vec<i32> = line.split(",").map(|n| n.parse().unwrap()).collect();
        if !u
            .windows(2)
            .any(|w| cache.get(&(w[0], w[1])).is_some_and(|v| !v))
        {
            out += u[u.len() / 2];
        }
    }
    out
}

fn part2() -> i32 {
    let (order, updates) = INPUT.split_once("\n\n").unwrap();

    let mut cache = HashMap::new();
    for line in order.lines() {
        let (x, y) = line
            .split_once("|")
            .map(|(x, y)| (x.parse::<i32>().unwrap(), y.parse::<i32>().unwrap()))
            .unwrap();
        cache.insert((x, y), -1);
        cache.insert((y, x), 1);
    }

    let mut out = 0;
    for line in updates.lines() {
        let mut u: Vec<i32> = line.split(",").map(|n| n.parse().unwrap()).collect();
        if u.windows(2)
            .any(|w| cache.get(&(w[0], w[1])).is_some_and(|&v| v == 1))
        {
            u.sort_by(|&x, &y| cache.get(&(x, y)).unwrap_or(&0).cmp(&0));
            out += u[u.len() / 2];
        }
    }
    out
}

fn main() {
    println!("part1 = {}", part1());
    println!("part2 = {}", part2());
}
